File: backend/utils/formatters.py
"""
Numeric formatting utilities for consistent decimal output.
Provides centralized rounding functions to ensure consistency across API responses.
"""
from typing import Optional

DECIMAL_STANDARD = 2  # Standard for prices, percentages
DECIMAL_PRECISION = 4  # For domain-specific values (correlations, factors)


def round2(value: Optional[float]) -> Optional[float]:
    """
    Round to 2 decimal places.
    Use for: prices, percentages, returns, scores.
    
    Args:
        value: Float value to round, or None
        
    Returns:
        Rounded float or None if input is None
    """
    if value is None:
        return None
    return round(value, DECIMAL_STANDARD)


def round4(value: Optional[float]) -> Optional[float]:
    """
    Round to 4 decimal places.
    Use for: correlation coefficients, factor exposure weights, NAV.
    
    Args:
        value: Float value to round, or None
        
    Returns:
        Rounded float or None if input is None
    """
    if value is None:
        return None
    return round(value, DECIMAL_PRECISION)


def round_percent(value: Optional[float]) -> Optional[float]:
    """
    Round percentage to 2 decimal places.
    Alias for round2() for semantic clarity.
    
    Args:
        value: Percentage value to round, or None
        
    Returns:
        Rounded percentage or None if input is None
    """
    return round2(value)


def round_price(value: Optional[float]) -> Optional[float]:
    """
    Round price to 2 decimal places.
    Alias for round2() for semantic clarity.
    
    Args:
        value: Price value to round, or None
        
    Returns:
        Rounded price or None if input is None
    """
    return round2(value)

File: frontend/src/utils/formatters.ts
/**
 * Numeric formatting utilities for consistent display.
 * Provides centralized formatting functions to replace duplicate implementations.
 */

export const DECIMAL_STANDARD = 2  // Standard for prices, percentages
export const DECIMAL_PRECISION = 4  // For correlations, factor exposure

/**
 * Format number with specified decimal places.
 * Handles null/undefined gracefully by returning '-'.
 */
export function formatNumber(
  val: number | null | undefined,
  decimals: number = DECIMAL_STANDARD,
  suffix: string = ''
): string {
  if (val === null || val === undefined || isNaN(val)) return '-'
  return `${val.toFixed(decimals)}${suffix}`
}

/**
 * Format percentage with 2 decimal places and % suffix.
 */
export function formatPercent(val: number | null | undefined): string {
  return formatNumber(val, DECIMAL_STANDARD, '%')
}

/**
 * Format price with 2 decimal places.
 */
export function formatPrice(val: number | null | undefined): string {
  return formatNumber(val, DECIMAL_STANDARD)
}

/**
 * Format number with sign (+/-) for changes.
 */
export function formatSign(
  val: number | null | undefined,
  suffix: string = ''
): string {
  if (val === null || val === undefined || isNaN(val)) return '-'
  const sign = val >= 0 ? '+' : ''
  return `${sign}${val.toFixed(DECIMAL_STANDARD)}${suffix}`
}

/**
 * Format percentage with sign (+/-).
 */
export function formatSignPercent(val: number | null | undefined): string {
  return formatSign(val, '%')
}

/**
 * Format correlation coefficient (4 decimal places).
 */
export function formatCorrelation(val: number | null | undefined): string {
  return formatNumber(val, DECIMAL_PRECISION)
}

/**
 * Format factor exposure weight (4 decimal places).
 */
export function formatFactorWeight(val: number | null | undefined): string {
  return formatNumber(val, DECIMAL_PRECISION)
}
